spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java

```diff
@@ -121,15 +137,18 @@ public Row next() {
       Row nextRow = rowIterator.next();
       cachedRow = nextRow;
 
-      if (isUpdateOrCarryoverRecord(currentRow, nextRow)) {
-        if (isCarryoverRecord(currentRow, nextRow)) {
-          // set carry-over rows to null for filtering out later
-          currentRow = null;
-          cachedRow = null;
-        } else {
-          currentRow = modify(currentRow, changeTypeIndex, UPDATE_BEFORE);
-          cachedRow = modify(nextRow, changeTypeIndex, UPDATE_AFTER);
-        }
+      if (sameLogicalRow(currentRow, nextRow)) {
+        String nextRowChangeType = nextRow.getString(changeTypeIndex);
+
+        Preconditions.checkState(
+            nextRowChangeType.equals(INSERT),
+            "The next row should be an INSERT row, but it is %s. That means there are multiple"
+                + " rows with the same value of identifier fields(%s). Please make sure the rows are unique.",
+            nextRowChangeType,
+            identifierFields);
```

Comment:
Is this array being nicely formatted? Can we add a test to verify the error message?

Reply:
Good catch. Fixed it.